File: frontend/src/geometry/coordinates.ts
import type { ABPoint, RotatedROI } from "../api/client";

export type SourceImageSize = {
  width: number;
  height: number;
};

export type DisplayRect = {
  width: number;
  height: number;
};

export type FrameDisplayTransform = {
  sourceWidth: number;
  sourceHeight: number;
  displayWidth: number;
  displayHeight: number;
  renderedWidth: number;
  renderedHeight: number;
  scale: number;
  offsetX: number;
  offsetY: number;
};

export function fitSourceToDisplay(
  source: SourceImageSize,
  display: DisplayRect
): FrameDisplayTransform {
  const scale = Math.min(display.width / source.width, display.height / source.height);
  const renderedWidth = source.width * scale;
  const renderedHeight = source.height * scale;
  return {
    sourceWidth: source.width,
    sourceHeight: source.height,
    displayWidth: display.width,
    displayHeight: display.height,
    renderedWidth,
    renderedHeight,
    scale,
    offsetX: (display.width - renderedWidth) / 2,
    offsetY: (display.height - renderedHeight) / 2
  };
}

export function displayPointToMeasurement(
  point: ABPoint,
  transform: FrameDisplayTransform,
  clamp = false
): ABPoint {
  const mapped = {
    x: (point.x - transform.offsetX) / transform.scale,
    y: (point.y - transform.offsetY) / transform.scale
  };
  if (!clamp) return mapped;
  return {
    x: clampValue(mapped.x, 0, transform.sourceWidth),
    y: clampValue(mapped.y, 0, transform.sourceHeight)
  };
}

export function measurementPointToDisplay(
  point: ABPoint,
  transform: FrameDisplayTransform
): ABPoint {
  return {
    x: point.x * transform.scale + transform.offsetX,
    y: point.y * transform.scale + transform.offsetY
  };
}

export function displayRoiToMeasurement(
  roi: RotatedROI,
  transform: FrameDisplayTransform
): RotatedROI {
  const center = displayPointToMeasurement({ x: roi.center_x, y: roi.center_y }, transform, true);
  return {
    ...roi,
    center_x: center.x,
    center_y: center.y,
    width: clampValue(roi.width / transform.scale, 1, transform.sourceWidth),
    height: clampValue(roi.height / transform.scale, 1, transform.sourceHeight)
  };
}

export function measurementRoiToDisplay(
  roi: RotatedROI,
  transform: FrameDisplayTransform
): RotatedROI {
  const center = measurementPointToDisplay({ x: roi.center_x, y: roi.center_y }, transform);
  return {
    ...roi,
    center_x: center.x,
    center_y: center.y,
    width: roi.width * transform.scale,
    height: roi.height * transform.scale
  };
}

export function roiCorners(roi: RotatedROI): ABPoint[] {
  const theta = (roi.angle_deg * Math.PI) / 180;
  const cos = Math.cos(theta);
  const sin = Math.sin(theta);
  return [
    [-roi.width / 2, -roi.height / 2],
    [roi.width / 2, -roi.height / 2],
    [roi.width / 2, roi.height / 2],
    [-roi.width / 2, roi.height / 2]
  ].map(([du, dv]) => ({
    x: roi.center_x + du * cos - dv * sin,
    y: roi.center_y + du * sin + dv * cos
  }));
}

function clampValue(value: number, low: number, high: number): number {
  return Math.max(low, Math.min(high, value));
}

File: backend/src/yyt1771_g3/core/coordinates.py
from __future__ import annotations

import math
from dataclasses import dataclass

from yyt1771_g3.core.models import ABPoint, RotatedROI


@dataclass(frozen=True)
class SourceImageSize:
    width: int
    height: int


@dataclass(frozen=True)
class DisplayRect:
    width: float
    height: float


@dataclass(frozen=True)
class FrameDisplayTransform:
    source_width: int
    source_height: int
    display_width: float
    display_height: float
    rendered_width: float
    rendered_height: float
    scale: float
    offset_x: float
    offset_y: float


def fit_source_to_display(source: SourceImageSize, display: DisplayRect) -> FrameDisplayTransform:
    if source.width <= 0 or source.height <= 0:
        raise ValueError("source image size must be positive")
    if display.width <= 0 or display.height <= 0:
        raise ValueError("display rect must be positive")

    scale = min(display.width / source.width, display.height / source.height)
    rendered_width = source.width * scale
    rendered_height = source.height * scale
    return FrameDisplayTransform(
        source_width=source.width,
        source_height=source.height,
        display_width=display.width,
        display_height=display.height,
        rendered_width=rendered_width,
        rendered_height=rendered_height,
        scale=scale,
        offset_x=(display.width - rendered_width) / 2.0,
        offset_y=(display.height - rendered_height) / 2.0,
    )


def display_point_to_measurement(
    x: float,
    y: float,
    transform: FrameDisplayTransform,
    *,
    clamp: bool = False,
) -> tuple[float, float]:
    mx = (x - transform.offset_x) / transform.scale
    my = (y - transform.offset_y) / transform.scale
    if clamp:
        mx = _clamp(mx, 0.0, float(transform.source_width))
        my = _clamp(my, 0.0, float(transform.source_height))
    return (mx, my)


def measurement_point_to_display(
    x: float,
    y: float,
    transform: FrameDisplayTransform,
) -> tuple[float, float]:
    return (
        x * transform.scale + transform.offset_x,
        y * transform.scale + transform.offset_y,
    )


def display_roi_to_measurement(
    roi: RotatedROI,
    transform: FrameDisplayTransform,
    *,
    clamp: bool = False,
) -> RotatedROI:
    center_x, center_y = display_point_to_measurement(
        roi.center_x,
        roi.center_y,
        transform,
        clamp=clamp,
    )
    width = roi.width / transform.scale
    height = roi.height / transform.scale
    if clamp:
        width = _clamp(width, 1.0, float(transform.source_width))
        height = _clamp(height, 1.0, float(transform.source_height))
    return RotatedROI(
        center_x=center_x,
        center_y=center_y,
        width=width,
        height=height,
        angle_deg=roi.angle_deg,
    )


def measurement_roi_to_display(roi: RotatedROI, transform: FrameDisplayTransform) -> RotatedROI:
    center_x, center_y = measurement_point_to_display(roi.center_x, roi.center_y, transform)
    return RotatedROI(
        center_x=center_x,
        center_y=center_y,
        width=roi.width * transform.scale,
        height=roi.height * transform.scale,
        angle_deg=roi.angle_deg,
    )


def roi_local_to_measurement_point(roi: RotatedROI, u: float, v: float) -> ABPoint:
    theta = math.radians(roi.angle_deg)
    du = u - roi.width / 2.0
    dv = v - roi.height / 2.0
    cos_t = math.cos(theta)
    sin_t = math.sin(theta)
    return ABPoint(
        x=roi.center_x + du * cos_t - dv * sin_t,
        y=roi.center_y + du * sin_t + dv * cos_t,
    )


def _clamp(value: float, low: float, high: float) -> float:
    return max(low, min(high, value))
